renku/core/management/template/template.py

```diff
@@ -0,0 +1,511 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Template management."""
+
+import json
+import re
+import shutil
+import tempfile
+from enum import Enum
+from functools import total_ordering
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Tuple
+
+from packaging.version import Version
+
+from renku.core import errors
+from renku.core.metadata.repository import Repository
+from renku.core.models.template import (
+    TEMPLATE_MANIFEST,
+    RenderedTemplate,
+    Template,
+    TemplateMetadata,
+    TemplateParameter,
+    TemplatesManifest,
+    TemplatesSource,
+)
+from renku.core.utils import communication
+from renku.core.utils.git import clone_repository
+from renku.core.utils.os import hash_file
+from renku.core.utils.util import to_semantic_version, to_string
+
+try:
+    import importlib_resources
+except ImportError:
+    import importlib.resources as importlib_resources
+
+TEMPLATE_KEEP_FILES = ["readme.md", "readme.rst", "readme.txt", "readme"]
+TEMPLATE_INIT_APPEND_FILES = [".gitignore"]
+
+
+class TemplateAction(Enum):
+    """Types of template rendering."""
+
+    INITIALIZE = 1
+    SET = 2
+    UPDATE = 3
+
+
+@total_ordering
+class FileAction(Enum):
+    """Types of operation when copying a template to a project."""
+
+    APPEND = 1
+    CREATE = 2
+    DELETED = 3
+    IGNORE_IDENTICAL = 4
+    IGNORE_UNCHANGED_REMOTE = 5
+    KEEP = 6
+    OVERWRITE = 7
+    RECREATE = 8
+
+    def __lt__(self, other):
+        if self.__class__ is other.__class__:
+            return self.value < other.value
+        return NotImplemented
```

Comment:
I didn't use `auto()` here because its docs says that the order is an implementation detail and might change in future. The order here is somewhat important since we show messages in that order.